cpp/src/join/join.cu

```diff
@@ -0,0 +1,551 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/copying.hpp>
+#include <cudf/join.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/detail/gather.hpp>
+#include <cudf/detail/gather.cuh>
+
+#include <join/hash_join.cuh>
+#include <algorithm>
+
+namespace cudf {
+
+namespace experimental {
+
+namespace detail {
+
+bool is_trivial_join(
+                     table_view const& left,
+                     table_view const& right,
+                     std::vector<size_type> const& left_on,
+                     std::vector<size_type> const& right_on,
+                     join_kind JoinKind) {
+  // If there is nothing to join, then send empty table with all columns
+  if (left_on.empty() || right_on.empty()) {
+      return true;
+  }
+
+  // Even though the resulting table might be empty, but the column should match the expected dtypes and other necessary information
```

Comment:
This comment is hard for me to parse and understand.

Reply:
This was a copy paste from legacy. To be frank I don't get it as well but I'll try to understand and reword it.

Reply:
The comment just meant that in the case where we return an empty table, the columns should still have the appropriate data types.

Reply:
I don't believe it is relevant to this function anymore.

Reply:
Removed